lizrd/core/llm.py

```diff
@@ -128,102 +130,78 @@ def LowRank(dinput, doutput, dlowrank):
     )
 
 
+def attention_mechanism(
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    dhead: int,
+    causal: bool,
+    flash: bool,
+):
+    if flash:
+        with torch.backends.cuda.sdp_kernel(
+            enable_flash=True, enable_math=False, enable_mem_efficient=False
+        ):
+            prefinal = F.scaled_dot_product_attention(
+                query=query,
+                key=key,
+                value=value,
+                attn_mask=None,
+                is_causal=causal,
+            )
+    else:
+        # implementation without flash assumes other dim order
+        query = query.transpose(1, 2)
+        key = key.transpose(1, 2)
+        value = value.transpose(1, 2)
+
+        a = torch.einsum("... l h d, ... L h d -> ... h l L", query, key)
+        a = a * (1 / dhead**0.5)
+        if causal:
+            a.masked_fill_(
+                torch.tril(torch.ones_like(a)) == 0, float("-inf")
+            )  # mask out future tokens
+        a = torch.softmax(a, dim=-1)
+        prefinal = torch.einsum("... h l L, ... L h d -> ... l h d", a, value)
+        prefinal = prefinal.transpose(1, 2)
+
+    return prefinal
```

Comment:
prefinal seems a bit nondescript, i propose `scaled_attention`  
```suggestion
def attention_mechanism(
    query: torch.Tensor,
    key: torch.Tensor,
    value: torch.Tensor,
    dhead: int,
    causal: bool,
    flash: bool,
):
    if flash:
        with torch.backends.cuda.sdp_kernel(
            enable_flash=True, enable_math=False, enable_mem_efficient=False
        ):
            scaled_attention = F.scaled_dot_product_attention(
                query=query,
                key=key,
                value=value,
                attn_mask=None,
                is_causal=causal,
            )
    else:
        # implementation without flash assumes other dim order
        query = query.transpose(1, 2)
        key = key.transpose(1, 2)
        value = value.transpose(1, 2)

        a = torch.einsum("... l h d, ... L h d -> ... h l L", query, key)
        a = a * (1 / dhead**0.5)
        if causal:
            a.masked_fill_(
                torch.tril(torch.ones_like(a)) == 0, float("-inf")
            )  # mask out future tokens
        a = torch.softmax(a, dim=-1)
        scaled_attention = torch.einsum("... h l L, ... L h d -> ... l h d", a, value)
        scaled_attention = prefinal.transpose(1, 2)

    return scaled_attention
```

Reply:
You are right, I just changed to output, imo makes sense (it's just the output of this function)